# LBA - A day in the life of Minervan Part II

You have 𝑛 activities that you would like to schedule in a day. These activities have all been preassigned with a
profit value, which you shall justify, and they need to be organized in an optimal scheduler. Here are a few
metrics your algorithm will need to take into account:

● The profit value measures the priority that a given activity should have. The higher the profit value the more
important it is to schedule that activity first.

● Time constraints restrict the number of activities that you are able to perform to, say, 𝑥 activities, where 𝑥 ≤ 𝑛.

● You will need to provide a short explanation about what it means to have an optimal scheduler. Perhaps if you
are planning several activities across town, you would like to save money on fuel/bus tickets, or perhaps you
would like to spend the least amount of time doing a number of tasks in a single day.

# 1. Analysis of Schedular 1.0

The simple task schedular was sorting through a number of tasks and scheduling tasks with fixed times at their given times, however it did not take into consideration the utility obtained from each task and hence was not an optimal algorithm for scheduling tasks. Its scheduling abilities were based on the dependencies, i.e., a task was performed if it has no dependencies and did it serially but following the order of the inputs.  This allowed for very restricted inputs that made the algorithm inflexible and inapplicable to different situations. We could not add a new task with high priority into the list and expect it to be scheduled accordingly. 
Moreover, since the algorithm didn’t schedule the tasks based on their utility, there was no way of expressing the significance of the task or the costs associated with it. The priority of the task should be dependent on not just our preferences but also real-world costs that can be expressed numerically and represented in our schedular. 
Lastly, it did not consider an upper bound (end of the day) and hence could not schedule tasks optimally within that bound. If we can only fit x number of tasks in a day, the schedular should choose the tasks that give the highest utility to be considered optimal. 






# 2. Algorithmic strategies: 

In [105]:
class MaxHeapq:
    """ 
    A class that implements properties and methods 
		that support a max priority queue data structure

		Attributes
	  ----------
	  heap : arr
	      A Python list where key values in the max heap are stored
	  heap_size: int
	      An integer counter of the number of keys present in the max heap
	  """  

    def __init__(self):       
        self.heap       = []
        self.heap_size  = 0
        
    def left(self, i):
       
        return 2 * i + 1

    def right(self, i):

        return 2 * i + 2

    def parent(self, i):
    

        return (i - 1)//2

    def maxk(self):     

        return self.heap[0]     
    
  
    def heappush(self, key):  
        self.heap.append(Task(-1, '', 10, [], -float("inf"), 0))
        self.increase_key(self.heap_size,key)
        self.heap_size+=1
        
    def increase_key(self, i, key): 
        if key.priority < self.heap[i].priority:
            raise ValueError('new key is smaller than the current key')
        self.heap[i] = key
        while i > 0 and self.heap[self.parent(i)].priority < self.heap[i].priority:
            j = self.parent(i)
            holder = self.heap[j]
            self.heap[j] = self.heap[i]
            self.heap[i] = holder
            i = j    
            
       
    def heapify(self, i):
        l = self.left(i)
        r = self.right(i)
        heap = self.heap
        if l <= (self.heap_size-1) and heap[l].priority >heap[i].priority:
            largest = l
        else:
            largest = i
        if r <= (self.heap_size-1) and heap[r].priority > heap[largest].priority:
            largest = r
        if largest != i:
            heap[i], heap[largest] = heap[largest], heap[i]
            self.heapify(largest)

    def heappop(self):
        if self.heap_size < 1:
            raise ValueError('Heap underflow: There are no keys in the priority queue ')
        maxk = self.heap[0]
        self.heap[0] = self.heap[-1]
        self.heap.pop()
        self.heap_size-=1
        self.heapify(0)
        return maxk

## Greedy Algorithm 

Description of the strategy

The prime feature aimed to be fulfilled is the greedy property in the greedy algorithmic approach, meaning a favorable variant will be made for the problem altogether, if the foremost options available are chosen at that moment. There is a possibility to attain an optimal outcome in some instances, despite the fact that this advancement is not a warranty for the finest outcomes for each trial. 
For this greedy algorithm, I have taken two approaches: 
The algorithm’s all-embracing approach considers the accessible problems with the highest weightage, then examines if they are feasible to be executed i.e for the time allocated, the upper limit is not surpassed and then, if it passes the checks the operation is completed. Hence, for every option, we make the best accessible option that is feasible while assuming the first task is not feasible, then approaching the next task with the highest weightage.


#### Pesudocode

Pesudocode

class Greedy_MultiTaskScheduler1:

def run_task_scheduler(self, starting_time = 480):

    current_time = starting_time       #set start time to be current time
    
    profit = 0                         #set total profit equal to 0
    
    get a list of fixed tasks
    
    while (there are unscheduled tasks) or (tasks in the priority queue):
    
        get a list of tasks without dependencies i.e. tasks whose dependencies have been done already
        
        if current_time//60 > 24:   #check to make sure day hasn't ended
        
            break #end the scheduler
            
        if tasks in priority queue > 1:
        
            if number of fixed tasks > than 1 and (time taken for the first activity in priority queue takes longer than time till fixed task):
           
                temporary = [] #empty list
                
                while (tasks in priority queue and tasks in priority queue cannot be done before the next fixed task):
                
                    add them to the temporary list
                    
                if tasks in priority queue:
                
                    task = root node of the priority queue
                    
                    current_time = self.multi_tasker(task, current_time) #call method to see if we can multitask with the task at hand
                    
                    profit = profit + task.priority #adding to total profit
                else:
                
                    current_time = current_time - duration of next fixed task
                    
                    task = first task in fixed task
                    
                    current_time = slef.multi_tasker(task, current_time)
                    
                    profit = profit + task.priority
                    
               for i in temporary:
               
                   push tasks into heap again
                   
           else:
           
               task = first task in priority queue
               
               current_time = self.multi_tasker(task, current_time)
               
               profit = profit + task.priority
               
       else:
       
           while tasks in fixed tasks heap:
           
               task = first tasks in the heap
               
               current_time = current_time + task.duration
               
               current_time = self.multi_tasker(task, current_time)
               
               profit = profit + task.priority
               
   total_time = current_time - starting_time

   print(time taken to complete the tasks)

   print(total profit from the activities)


#### Data Strcutures uses:
The data structures were used in the dynamic programming task scheduler are heaps, classes, and tables.

Heaps are implemented to sort the tasks based on the priority that they have. This allows us to sort them easily as heaps are an effective and efficient way of sorting tasks as compared methods such as like sorting. They also makes retrieval of tasks much easier as we just need to extract the root node in order to get the task with the highest priority.

Classes: three classes have been implemented.

MaxHeap: class that has the attributes and methods needed for the heap.
Task class: takes the list input and compiles them into a formate we can work with in our schedular. It only has attributes and no functionality as it is just used for converting the input into the required formate.
TaskScheduler class: has the main functionality of the algorithm in its methods.
The Task and the Task Schedular class created separately because they offer easy maintenance of the code. Changes to the TaskScheduler class do not affect the way the input is given and the Task class makes it easier to change the way a certain input is percieved by the algorithm which was very useful especially when two different implementations of the greedy algorithm approach were done as it allowed versatility of approaches with respect to the utility value.


#### Assumptions:

The assumption we make here is that the user of the algorithm has knowledge of the perceived costs and benefits associated with each activity. Since this is the overarching fundamental concept, without this, the algorithm is useless. If the user is unable to comprehend this principle, they might end up performing tasks that could be highly inefficient, have low aggregate profit values, or processes that are extremely costly.

The second assumption is that the user understands the limited nature of the resource of time, and knows that some tasks might exceed the set threshold and the tasks might be ignored or removed. The user needs to take note of this and take measures to ensure that any task missed by the algorithm is in the users' knowledge.  For example, if a compulsory class is being skipped by the algorithm and user takes note of it then they can go back and tweak the profits and costs associated.

In [106]:
#code for Greedy algorithm schdular 

#test case

In [107]:
class Task_1:
    """
    - id: Task Id   
    - description: Short description of the task   
    - duration: Duration in minutes   
    - dependencies: An array of id's of tasks that the task depends on 
    - status: Current status of the task
    - multi_tasking: Tells whether the task can be multitasked or not
    - fixed: Bool for if the task has a fixed start time or can start at any time within its urgency
    - priority: The urgency of the task 
   
    """
    #Initializes an instance of Task, since I am using the same code for Q4, I just set the multitasking part to be
    #default false in this case so that we do not need to account for it in this question.
    #the default priority I am using is 1440 which is equal to the number of minutes in a day
    #fixed is the attribute used to account for the case that there is a task that needs to be done at a certain time
    #in the day
    def __init__(self,task_id,description,duration,dependencies, priority, cost = 0, fixed=0, multi_tasking=False, status="N"):
        self.id = task_id
        self.description = description
        self.duration = duration
        self.dependencies = dependencies
        self.status = status
        self.priority = round(((priority - cost)/duration), 5)
        self.multi_tasking = multi_tasking
        self.fixed = fixed
    
    #The format of the display
    def __repr__(self):
        return f"{self.description} - id: {self.id}\n \tDuration:{self.duration}\n\tDepends on: {self.dependencies}\n\tMust be completed at most {self.priority} minutes after day start\n\tFixed task start: {self.fixed}\n\tStatus: {self.status}"
    
    #Definining the comparison variable which in this case is the priority
    def __lt__(self, other):
        return self.priority < other.priority


In [142]:

class Greedy_MultiTaskScheduler1:
    """
    A Daily Task Scheduler Using Priority Queues based on priority that allows for multi-tasking
    Attributes
    ----------
    tasks : arr
        A Python list of all the tasks we schedule
    priority_queue: arr
        The priority queue of tasks ready to execute
    fixed_tasks: arr
        The priority queue of tasks that start at a fixed time
    """ 
    
    NOT_STARTED ='N'
    IN_PRIORITY_QUEUE = 'I'
    COMPLETED = 'C'
    
    def __init__(self, tasks):
        self.tasks = tasks
        self.priority_queue = MaxHeapq()
        self.fixed_tasks = MaxHeapq()
                
    def print_self(self):
        print('Input List of Tasks')
        for t in self.tasks:
            print(t)            
            
    def remove_dependency(self, task_id):
        """
        Input: list of tasks and id of the task completed
        Output: lists of tasks with completed task removed
        """
        for t in self.tasks:
            if t.id != task_id and task_id in t.dependencies:
                t.dependencies.remove(task_id)           
            
    def get_tasks_ready(self):
        """ 
        Input: list of tasks
        Output: list of tasks that are executable right now(no dependencies)
        """
        for task in self.tasks:
            if task.status == self.NOT_STARTED and not task.dependencies and not task.fixed: 
                task.status = self.IN_PRIORITY_QUEUE 
                self.priority_queue.heappush(task)
    
#    def update_priority(self, time_elapsed=0):
        """
        Input: duration of the most recent executed task
        Output: boolean (checks the status of all tasks and returns True if at least one task has status = 'N'
        """
#        for i in range(len(self.tasks)): 
#            self.tasks[i].priority -= time_elapsed
        
    def check_unscheduled_tasks(self):
        """
        Input: list of tasks 
        Output: boolean (checks the status of all tasks and returns True if at least one task has status = 'N'
        """
        for task in self.tasks:
            if task.status == self.NOT_STARTED:
                return True
        return False   
    
    def get_fixed_tasks(self):
        """
        Get all tasks in the task list that have fixed start times
        """
        for i in self.tasks:
            if i.fixed:
                self.fixed_tasks.heappush(i)
    
    def format_time(self, time):
        return f"{time//60}h{time%60:02d}"
    
    #method that checks for and executes multi-tasking
    def multi_tasker(self, task, current_time, profit):
        if (task.multi_tasking):
            temporary = []
            while(self.priority_queue.heap_size > 0 and (not self.priority_queue.maxk().multi_tasking or (self.fixed_tasks.heap_size > 0 and (self.priority_queue.maxk().duration > self.fixed_tasks.maxk().duration)))):
                temporary.append(self.priority_queue.heappop())

# if there are tasks in the priority queue after the while loop is done, we can multi-task using them
            if self.priority_queue.heap:
                multitask = self.priority_queue.heappop()
                if ((current_time + task.duration)//60 < 24) and ((current_time + multitask.duration)//60 < 24):
                    print(f"⏰Simple Scheduler at time {self.format_time(current_time)} started executing tasks {task.id} and {multitask.id} that take {task.duration} and {multitask.duration} mins respectively")
                
                #Increment the current time with the duration of the larger task(duration wise) as that task will still
                #be in process when the first task is completed.
                if task.duration < multitask.duration:
                    current_time += task.duration
                    if ((current_time + task.duration)//60 < 24) and ((current_time + multitask.duration)//60 < 24):
                        print(f"✅ Completed Task {task.id} - '{task.description}' at time {self.format_time(current_time)}\n")
                        current_time += multitask.duration - task.duration
                        print(f"✅ Completed Task {multitask.id} - '{multitask.description}' at time {self.format_time(current_time)}\n")
                        #self.update_priority(multitask.duration)
                        profit += (task.priority*task.duration) + (multitask.priority*task.duration)
                        
        
                else:
                    current_time += multitask.duration
                    if ((current_time + task.duration)//60 < 24) and ((current_time + multitask.duration)//60 < 24):
                        print(f"✅ Completed Task {multitask.id} - '{multitask.description}' at time {self.format_time(current_time)}\n")
                        current_time += task.duration - multitask.duration
                        print(f"✅ Completed Task {task.id} - '{task.description}' at time {self.format_time(current_time)}\n")
                        #self.update_priority(task.duration)
                        profit += (task.priority*task.duration) + (multitask.priority*task.duration)
                              
                
                self.remove_dependency(task.id)
                self.remove_dependency(multitask.id)
                task.status = self.COMPLETED
                multitask.status = self.COMPLETED

            #if the priority queue is empty after the while loop:
            #none of the tasks can be multitasked 
            #within the time between the current time and the start of the next fixed task
            #so we just perform the initial task without multi-tasking
            else:
                if ((current_time + task.duration)//60 < 24):
                    print(f"⏰Simple Scheduler at time {self.format_time(current_time)} started executing task {task.id} that takes {task.duration} mins")
                    current_time += task.duration            
                    print(f"✅ Completed Task {task.id} - '{task.description}' at time {self.format_time(current_time)}\n")  
                    #self.update_priority(task.duration)
                    profit += (task.priority*task.duration)
                    
                self.remove_dependency(task.id)
                task.status = self.COMPLETED
                

#once we've carried out the task:
            #we move all the other tasks to the priority queue from the temporary list we had added them to
            for i in temporary:
                self.priority_queue.heappush(i)
            
        #if the task isn't multi-taskable we treat it as a regular task without running it through further comparisons
        else: 
            if (current_time + task.duration)//60 < 24:
                print(f"⏰Simple Scheduler at time {self.format_time(current_time)} started executing task {task.id} that takes {task.duration} mins")
                current_time += task.duration            
                print(f"✅ Completed Task {task.id} - '{task.description}' at time {self.format_time(current_time)}\n")
                profit += (task.priority*task.duration)
                

            #self.update_priority(task.duration)
            self.remove_dependency(task.id)
            task.status = self.COMPLETED
            
        #return the current time
        return current_time, profit
    
    def run_task_scheduler(self, starting_time = 480):
        current_time = starting_time
        profit = 0
        self.get_fixed_tasks()
        while self.check_unscheduled_tasks() or self.priority_queue.heap:
            self.get_tasks_ready()
            if current_time//60 > 24:
                print("End of the day. No more tasks possible today")
                break
            if self.priority_queue.heap_size > 0:      
                if self.fixed_tasks.heap_size > 0 and self.priority_queue.maxk().duration > self.fixed_tasks.maxk().duration:  
                    temporary = []
                    while(self.priority_queue.heap_size > 0 and self.priority_queue.maxk().duration > self.fixed_tasks.maxk().duration):
                        temporary.append(self.priority_queue.heappop())   
                    if self.priority_queue.heap:
                        task = self.priority_queue.heappop()  
                        
                        # calling the multi_task method and update the current time to the 
                        #new time once the multi-task method is done.
                        current_time, profit = self.multi_tasker(task, current_time, profit)
                        #profit += task.priority
                        print(profit)

                    else:
                     
                        #current_time = current_time - self.fixed_tasks.maxk().duration
                        #self.update_priority(self.fixed_tasks.mink().priority - self.fixed_tasks.mink().duration)
                        task = self.fixed_tasks.heappop()
                        
                        current_time, profit = self.multi_tasker(task, current_time, profit)
                        #profit += task.priority
                       
                    for i in temporary:
                        self.priority_queue.heappush(i)
                        
                #if the current task can be done before the next fixed task but can not be multi-tasked with another
                #task, we just perform it like a normal task.
                else:
                    task = self.priority_queue.heappop()   
                    
                    current_time, profit = self.multi_tasker(task, current_time, profit)
                    #profit += task.priority
                    
            else:
                while self.fixed_tasks.heap_size > 0:
                    current_time = current_time - self.fixed_tasks.maxk().duration
                    #self.update_priority(self.fixed_tasks.mink().priority - self.fixed_tasks.mink().duration)
                    task = self.fixed_tasks.heappop()
                    
                    current_time, profit = self.multi_tasker(task, current_time, profit)
                    #profit += task.priority
                     
        total_time = current_time - starting_time             
        print(f"🏁 Completed all planned tasks in {total_time//60}h{total_time%60:02d}min")
        print(f"Total profit = {profit}")

In [161]:
#test cases for greedy algorithm 1:
task_list = [
    Task_1(0, 'waking up', 20, [], 100, 0, fixed = True),
    Task_1(1, 'having breakfast', 20, [0], 75, multi_tasking=True),
    Task_1(2, 'watch an episode of Community', 30, [], 35, 5, multi_tasking = True),
    Task_1(3, 'take a walk', 30, [], 25, 10, multi_tasking = True),
    Task_1(4, 'Do pre-class work', 180, [] , 80),
    Task_1(5, 'Take class', 90, [4], 90),
    Task_1(6, 'Go to a park', 180, [], 50, 5, multi_tasking = True),
    Task_1(7, 'take a trip to the beach', 120, [], 60, 20),
    Task_1(8, 'get groceries', 50, [], 70, 30),
    Task_1(9, 'meal prep', 40, [8], 50, 10),
    Task_1(10, 'socialize', 200, [], 60, 5)]

task_scheduler = Greedy_MultiTaskScheduler1(task_list)
task_scheduler.run_task_scheduler()


⏰Simple Scheduler at time 8h00 started executing task 0 that takes 20 mins
✅ Completed Task 0 - 'waking up' at time 8h20

⏰Simple Scheduler at time 8h20 started executing tasks 1 and 2 that take 20 and 30 mins respectively
✅ Completed Task 1 - 'having breakfast' at time 8h40

✅ Completed Task 2 - 'watch an episode of Community' at time 8h50

⏰Simple Scheduler at time 8h50 started executing task 8 that takes 50 mins
✅ Completed Task 8 - 'get groceries' at time 9h40

⏰Simple Scheduler at time 9h40 started executing task 9 that takes 40 mins
✅ Completed Task 9 - 'meal prep' at time 10h20

⏰Simple Scheduler at time 10h20 started executing tasks 3 and 6 that take 30 and 180 mins respectively
✅ Completed Task 3 - 'take a walk' at time 10h50

✅ Completed Task 6 - 'Go to a park' at time 13h20

⏰Simple Scheduler at time 13h20 started executing task 4 that takes 180 mins
✅ Completed Task 4 - 'Do pre-class work' at time 16h20

⏰Simple Scheduler at time 16h20 started executing task 5 that takes 90

In [110]:
class Task:
    """
    - id: Task Id   
    - description: Short description of the task   
    - duration: Duration in minutes   
    - dependencies: An array of id's of tasks that the task depends on 
    - status: Current status of the task
    - multi_tasking: Tells whether the task can be multitasked or not
    - fixed: Bool for if the task has a fixed start time or can start at any time within its urgency
    - priority: The urgency of the task 
   
    """
    #Initializes an instance of Task, since I am using the same code for Q4, I just set the multitasking part to be
    #default false in this case so that we do not need to account for it in this question.
    #the default priority I am using is 1440 which is equal to the number of minutes in a day
    #fixed is the attribute used to account for the case that there is a task that needs to be done at a certain time
    #in the day
    def __init__(self,task_id,description,duration,dependencies, priority, cost = 0, fixed=False, multi_tasking=False, status="N"):
        self.id = task_id
        self.description = description
        self.duration = duration
        self.dependencies = dependencies
        self.status = status
        self.cost = cost
        self.priority = priority - cost
        self.multi_tasking = multi_tasking
        self.fixed = fixed
    
    #The format of the display
    def __repr__(self):
        return f"{self.description} - id: {self.id}\n \tDuration:{self.duration}\n\tDepends on: {self.dependencies}\n\tMust be completed at most {self.priority} minutes after day start\n\tFixed task start: {self.fixed}\n\tStatus: {self.status}"
    
    #Definining the comparison variable which in this case is the priority
    def __lt__(self, other):
        return self.priority < other.priority



In [135]:

class Greedy_MultiTaskScheduler2:
    """
    A Daily Task Scheduler Using Priority Queues based on priority that allows for multi-tasking
    Attributes
    ----------
    tasks : arr
        A Python list of all the tasks we schedule
    priority_queue: arr
        The priority queue of tasks ready to execute
    fixed_tasks: arr
        The priority queue of tasks that start at a fixed time
    """ 
    
    NOT_STARTED ='N'
    IN_PRIORITY_QUEUE = 'I'
    COMPLETED = 'C'
    
    def __init__(self, tasks):
        self.tasks = tasks
        self.priority_queue = MaxHeapq()
        self.fixed_tasks = MaxHeapq()
                
    def print_self(self):
        print('Input List of Tasks')
        for t in self.tasks:
            print(t)            
            
    def remove_dependency(self, task_id):
        """
        Input: list of tasks and id of the task completed
        Output: lists of tasks with completed task removed
        """
        for t in self.tasks:
            if t.id != task_id and task_id in t.dependencies:
                t.dependencies.remove(task_id)           
            
    def get_tasks_ready(self):
        """ 
        Input: list of tasks
        Output: list of tasks that are executable right now(no dependencies)
        """
        for task in self.tasks:
            if task.status == self.NOT_STARTED and not task.dependencies and not task.fixed: 
                task.status = self.IN_PRIORITY_QUEUE 
                self.priority_queue.heappush(task)
    
#    def update_priority(self, time_elapsed=0):
        """
        Input: duration of the most recent executed task
        Output: boolean (checks the status of all tasks and returns True if at least one task has status = 'N'
        """
#        for i in range(len(self.tasks)): 
#            self.tasks[i].priority -= time_elapsed
        
    def check_unscheduled_tasks(self):
        """
        Input: list of tasks 
        Output: boolean (checks the status of all tasks and returns True if at least one task has status = 'N'
        """
        for task in self.tasks:
            if task.status == self.NOT_STARTED:
                return True
        return False   
    
    def get_fixed_tasks(self):
        """
        Get all tasks in the task list that have fixed start times
        """
        for i in self.tasks:
            if i.fixed:
                self.fixed_tasks.heappush(i)
    
    def format_time(self, time):
        return f"{time//60}h{time%60:02d}"
    
    #method that checks for and executes multi-tasking
    def multi_tasker(self, task, current_time, profit):
        if (task.multi_tasking):
            temporary = []
            while(self.priority_queue.heap_size > 0 and (not self.priority_queue.maxk().multi_tasking or (self.fixed_tasks.heap_size > 0 and (self.priority_queue.maxk().duration > self.fixed_tasks.maxk().duration)))):
                temporary.append(self.priority_queue.heappop())

# if there are tasks in the priority queue after the while loop is done, we can multi-task using them
            if self.priority_queue.heap:
                multitask = self.priority_queue.heappop()
                if ((current_time + task.duration)//60 < 24) and ((current_time + multitask.duration)//60 < 24):
                    print(f"⏰Simple Scheduler at time {self.format_time(current_time)} started executing tasks {task.id} and {multitask.id} that take {task.duration} and {multitask.duration} mins respectively")
                
                #Increment the current time with the duration of the larger task(duration wise) as that task will still
                #be in process when the first task is completed.
                if task.duration < multitask.duration:
                    current_time += task.duration
                    if ((current_time + task.duration)//60 < 24) and ((current_time + multitask.duration)//60 < 24):
                        print(f"✅ Completed Task {task.id} - '{task.description}' at time {self.format_time(current_time)}\n")
                        current_time += multitask.duration - task.duration
                        print(f"✅ Completed Task {multitask.id} - '{multitask.description}' at time {self.format_time(current_time)}\n")
                        #self.update_priority(multitask.duration)
                        profit += task.priority + multitask.priority
                        
        
                else:
                    current_time += multitask.duration
                    if ((current_time + task.duration)//60 < 24) and ((current_time + multitask.duration)//60 < 24):
                        print(f"✅ Completed Task {multitask.id} - '{multitask.description}' at time {self.format_time(current_time)}\n")
                        current_time += task.duration - multitask.duration
                        print(f"✅ Completed Task {task.id} - '{task.description}' at time {self.format_time(current_time)}\n")
                        #self.update_priority(task.duration)
                        profit += task.priority + multitask.priority
                              
                
                self.remove_dependency(task.id)
                self.remove_dependency(multitask.id)
                task.status = self.COMPLETED
                multitask.status = self.COMPLETED

            #if the priority queue is empty after the while loop:
            #none of the tasks can be multitasked 
            #within the time between the current time and the start of the next fixed task
            #so we just perform the initial task without multi-tasking
            else:
                if ((current_time + task.duration)//60 < 24):
                    print(f"⏰Simple Scheduler at time {self.format_time(current_time)} started executing task {task.id} that takes {task.duration} mins")
                    current_time += task.duration            
                    print(f"✅ Completed Task {task.id} - '{task.description}' at time {self.format_time(current_time)}\n")  
                    #self.update_priority(task.duration)
                    profit += task.priority
                    
                self.remove_dependency(task.id)
                task.status = self.COMPLETED
                

#once we've carried out the task:
            #we move all the other tasks to the priority queue from the temporary list we had added them to
            for i in temporary:
                self.priority_queue.heappush(i)
            
        #if the task isn't multi-taskable we treat it as a regular task without running it through further comparisons
        else: 
            if (current_time + task.duration)//60 < 24:
                print(f"⏰Simple Scheduler at time {self.format_time(current_time)} started executing task {task.id} that takes {task.duration} mins")
                current_time += task.duration            
                print(f"✅ Completed Task {task.id} - '{task.description}' at time {self.format_time(current_time)}\n")
                profit += task.priority
                

            #self.update_priority(task.duration)
            self.remove_dependency(task.id)
            task.status = self.COMPLETED
            
        #return the current time
        return current_time, profit
    
    def run_task_scheduler(self, starting_time = 480):
        current_time = starting_time
        profit = 0
        self.get_fixed_tasks()
        while self.check_unscheduled_tasks() or self.priority_queue.heap:
            self.get_tasks_ready()
            if current_time//60 > 24:
                print("End of the day. No more tasks possible today")
                break
            if self.priority_queue.heap_size > 0:      
                if self.fixed_tasks.heap_size > 0 and self.priority_queue.maxk().duration > self.fixed_tasks.maxk().duration:  
                    temporary = []
                    while(self.priority_queue.heap_size > 0 and self.priority_queue.maxk().duration > self.fixed_tasks.maxk().duration):
                        temporary.append(self.priority_queue.heappop())   
                    if self.priority_queue.heap:
                        task = self.priority_queue.heappop()  
                        
                        # calling the multi_task method and update the current time to the 
                        #new time once the multi-task method is done.
                        current_time, profit = self.multi_tasker(task, current_time, profit)
                        #profit += task.priority
                        print(profit)

                    else:
                     
                        #current_time = current_time - self.fixed_tasks.maxk().duration
                        #self.update_priority(self.fixed_tasks.mink().priority - self.fixed_tasks.mink().duration)
                        task = self.fixed_tasks.heappop()
                        
                        current_time, profit = self.multi_tasker(task, current_time, profit)
                        #profit += task.priority
                       
                    for i in temporary:
                        self.priority_queue.heappush(i)
                        
                #if the current task can be done before the next fixed task but can not be multi-tasked with another
                #task, we just perform it like a normal task.
                else:
                    task = self.priority_queue.heappop()   
                    
                    current_time, profit = self.multi_tasker(task, current_time, profit)
                    #profit += task.priority
                    
            else:
                while self.fixed_tasks.heap_size > 0:
                    current_time = current_time - self.fixed_tasks.maxk().duration
                    #self.update_priority(self.fixed_tasks.mink().priority - self.fixed_tasks.mink().duration)
                    task = self.fixed_tasks.heappop()
                    
                    current_time, profit = self.multi_tasker(task, current_time, profit)
                    #profit += task.priority
                     
        total_time = current_time - starting_time             
        print(f"🏁 Completed all planned tasks in {total_time//60}h{total_time%60:02d}min")
        print(f"Total profit = {profit}")

In [159]:
#test cases for greedy algorithm 2:
#some attributes of the class are taken as defaults, such as fixed, multitasking and cost. 
#If a different input is needed,then it is specified 
task_list = [
    Task(0, 'waking up', 20, [], 100, fixed = True),
    Task(1, 'having breakfast', 20, [0], 75, multi_tasking=True),
    Task(2, 'watch an episode of Community', 30, [], 35, 5, multi_tasking = True),
    Task(3, 'take a walk', 30, [], 25, 10, multi_tasking = True),
    Task(4, 'Do pre-class work', 180, [] , 80),
    Task(5, 'Take class', 90, [4], 90),
    Task(6, 'Go to a park', 180, [], 50, 5, multi_tasking = True),
    Task(7, 'take a trip to the beach', 120, [], 60, 20),
    Task(8, 'get groceries', 50, [], 70, 30),
    Task(9, 'meal prep', 40, [8], 50, 10),
    Task(10, 'socialize', 200, [], 60, 5)]

task_scheduler = Greedy_MultiTaskScheduler2(task_list)
task_scheduler.run_task_scheduler()

⏰Simple Scheduler at time 8h00 started executing task 0 that takes 20 mins
✅ Completed Task 0 - 'waking up' at time 8h20

⏰Simple Scheduler at time 8h20 started executing task 4 that takes 180 mins
✅ Completed Task 4 - 'Do pre-class work' at time 11h20

⏰Simple Scheduler at time 11h20 started executing task 5 that takes 90 mins
✅ Completed Task 5 - 'Take class' at time 12h50

⏰Simple Scheduler at time 12h50 started executing tasks 1 and 6 that take 20 and 180 mins respectively
✅ Completed Task 1 - 'having breakfast' at time 13h10

✅ Completed Task 6 - 'Go to a park' at time 15h50

⏰Simple Scheduler at time 15h50 started executing task 10 that takes 200 mins
✅ Completed Task 10 - 'socialize' at time 19h10

⏰Simple Scheduler at time 19h10 started executing task 7 that takes 120 mins
✅ Completed Task 7 - 'take a trip to the beach' at time 21h10

⏰Simple Scheduler at time 21h10 started executing task 8 that takes 50 mins
✅ Completed Task 8 - 'get groceries' at time 22h00

⏰Simple Scheduler

#### test case

The case shows that the schedular can effectively deal with fixed and multi-taskable tasks and schedules them accoridng to their priority. These priorities are calculated by doing a cost-benefit analysis (profit - cost) and calculating it either as a whole number or as priority/minute. These priority values allows it to make a choice for the best possible solution by selecting the task with the highest utility value, without taking into conisderation all the possible combinations. 
The schedular works as expected and also shows the corresponding profit value

In [ ]:
#example number 1

task_list_a = [
    Task_1(0, 'A', 100, [], 100),
    Task_1(1, 'B', 20, [0], 75),
    Task_1(2, 'C', 300, [], 35, 5),
    Task_1(3, 'D', 30, [], 25, 10),
    Task_1(4, 'E', 180, [] , 80),
    Task_1(5, 'F', 90, [4], 90),
    Task_1(6, 'G', 180, [], 50, 5),
    Task_1(7, 'H', 70, [], 60, 30), 
    Task_1(8, 'I', 40, [], 30, 10), 
    Task_1(7, 'J', 50, [], 70, 20)]


In [162]:
#two test cases 

task_list1 = [
    Task_1(0, 'A', 100, [], 100),
    Task_1(1, 'B', 20, [0], 75),
    Task_1(2, 'C', 300, [], 35, 5),
    Task_1(3, 'D', 30, [], 25, 10),
    Task_1(4, 'E', 180, [] , 80),
    Task_1(5, 'F', 90, [4], 90),
    Task_1(6, 'G', 180, [], 50, 5),
    Task_1(7, 'H', 70, [], 60, 30), 
    Task_1(8, 'I', 40, [], 30, 10), 
    Task_1(9, 'J', 50, [], 70, 20)]

task_list2 = [
    Task(0, 'A', 100, [], 100),
    Task(1, 'B', 20, [0], 75),
    Task(2, 'C', 300, [], 35, 5),
    Task(3, 'D', 30, [], 25, 10),
    Task(4, 'E', 180, [] , 80),
    Task(5, 'F', 90, [4], 90),
    Task(6, 'G', 180, [], 50, 5),
    Task(7, 'H', 70, [], 60, 30),
    Task(8, 'I', 40, [], 30, 10), 
    Task(9, 'J', 50, [], 70, 20)]


print("This is an example of the first greedy task schedular with utility/minute")
task_scheduler1 = Greedy_MultiTaskScheduler1(task_list1)
task_scheduler1.run_task_scheduler()
print('----------------------------------------------------------------------------')
print("This is an example of the Second greedy task schedular with whole utilities")
task_scheduler2 = Greedy_MultiTaskScheduler2(task_list2)
task_scheduler2.run_task_scheduler()


This is an example of the first greedy task schedular with utility/minute
⏰Simple Scheduler at time 8h00 started executing task 0 that takes 100 mins
✅ Completed Task 0 - 'A' at time 9h40

⏰Simple Scheduler at time 9h40 started executing task 1 that takes 20 mins
✅ Completed Task 1 - 'B' at time 10h00

⏰Simple Scheduler at time 10h00 started executing task 9 that takes 50 mins
✅ Completed Task 9 - 'J' at time 10h50

⏰Simple Scheduler at time 10h50 started executing task 8 that takes 40 mins
✅ Completed Task 8 - 'I' at time 11h30

⏰Simple Scheduler at time 11h30 started executing task 3 that takes 30 mins
✅ Completed Task 3 - 'D' at time 12h00

⏰Simple Scheduler at time 12h00 started executing task 4 that takes 180 mins
✅ Completed Task 4 - 'E' at time 15h00

⏰Simple Scheduler at time 15h00 started executing task 5 that takes 90 mins
✅ Completed Task 5 - 'F' at time 16h30

⏰Simple Scheduler at time 16h30 started executing task 7 that takes 70 mins
✅ Completed Task 7 - 'H' at time 17h40


#### Analysis of the test case

1. The above list of tasks are an example of how the the two greedy algorithms can run differently depending on the input that is given. Our initial test cases for the two, with the schedule of a day as an input, showed how using the integer-utility value for priority calculations gave a combination with higher aggregate utility, this test shows the opposite. That is because both of the greedy approaches are valid and aim for an optimal schedular, but one algorithm can perform better for particular inputs as compared to others. Here, using utility/minute allowed the schedular to choose tasks that took less time and hence could fit 9 tasks in a day as compared to 8 that using integral utilities could. 
This input measured just the profit calculation of the two, without consideration of whether tasks could be multi-tasked or fixed in order to show a comparision of just how the optimal profit can change with a different input. 


2. Another example of how the schedular works as expected is that is takes into consideration an upper bound, a value of n within which all possible tasks can be done. This value of n can be specificed by the user given how much time they have in a day to schedule their tasks. I used a baseline value corresponding to the end of the day (12:00 am) as my value of n. 

### Time and Space Complexity

##### Time Complexity

Time Complexity


Time recurrence calculation

Costs the algorithm

Multi_tasker() = O(n) x O(n - 1)
the function runa on all the tasks once and then again without the task that was just performed. 
simplifies down to O(n^2) - O(n).

Run_taskscheduler() = O(n) x O(n - 1) 
runs the task and then runs the method again on the tasks that are left which also simplifies down to O(n^2) - O(n).

The methods, like get_tasks_ready(), and get_unscheduled_tasks() take O(n) time which repeated for a constant number of times will lead to C*O(n)

The runtime of the heappush() method = O(n log n) 

The extraction of the root nodes for the sake of tasks take O(1) time and multiplied by a constant means the total time will be K*O(1)

Thus, the time recurrence equation becomes:

T(n) = O(n^2) - O(n) + O(n^2) - O(n) + CO(n) + O(n log n) + KO(1)

The 2O(n) + CO(n) will become C*O(n) as C is just a constant

T(n) = 2 O(n^2) + O(n log n) + CO(n) + KO(1)

For a very large input size, only the O(n^2) will be important and the others can be disregarded as their growth rate is too small for significant impact.

Thus, the final time recurrence equation becomes:

T(n) = O(n^2)

This means that each task added to the scheduler will result in a n^2 factor increase in run time because when we add a task, the program will have an n^2 number of increase in the comparisons/iterations that it has to do over the task.

##### Space Complexity 

As greedy algorithm does not use a table like dynamic algorithm, it will take less space in comparison. his algorithm relies on the choices it makes at each possible decision point to guide it towards the optimal schedule.

In terms of numbers, the greedy algorithm takes up 4*n^2 bytes as each integer takes 4 bytes and there are n rows and columns in the table

## Dynamic Programming

Description of the strategy

Dynamic programming is an algorithm that aims to compute an optimal output by combining the solution of smaller subproblems and finding the optimal combination. We can use this algorithm to create an optimal schedular by computing the best possible combination that gives us the highest utility/profit value. This is possible by comparing different combinations and their respective profit values and then giving the user the best possible case. 

This approach uses a table to keep track of the all the profit values and then calculating the highest outcome. The algorithm runs over each possibility, comparing each profit value with the previous one. If it is higher than the previous, it puts it in the table. If it is lower, it takesn the previous value in that position too. Hence, when it reaches the end of the table, the table[n][m] value is the highest possible value, giving us the final output that we use.

#### Pesudocode

Pesudocode
class DP_MultiTaskScheduler:

def get_combinations(self):

    times = []
    
    profits = []
    
    for task in tasks:
    
        add task duration to times
        
        add task priority to profits
        
    return times, profits
    
    
def best_combinations(self, max time):

    times = first element of the tuple returned from combinations
    
    profits = second element of the tuple returned from combinations
    
    n = len(profits)
    
    K(initiate table) = [[for x from 0 to time] for x upto n + 1]
    
    for i from 1 to n:
    
        for w from 1 to time:
        
            if i or w are 1:
            
                set corresponding table index to 0
                
            else if times[i - 1] < = w:
            
                set corresponding table index to be the higher value from current and the index before it
                
            else:
            
                set corresponding table index to have an equal value to the last visited position in the table
    return K[n][w]


def run_task_scheduler(self, starting time):

    max profit = self.best_possible_combinations()
   
    current_time = starting_time
    
    profit = 0
    
    n = the total number of hours to run the schedular for
    
    while (there are unscheduled tasks) or (tasks in priority queue):
    
        get a list of tasks that can be done right now
        
        if current_time > n:
        
            break
            
        if tasks in the priority queue:
        
            temporary = []
            
            add all tasks
            
            take the first task from the possible ones
            
            update current time and priority with multi_tasker(task, current_time, profit)
            
            for tasks in priority:
            
                push them into the heap again
                
        else:
        
            task = first task in the priority_queue
            
            update current time and priority with multi_tasker(task, current_time, profit)
            
    total_time = current_time - starting_time
    
    print current time and profit achieved
    
    
#### Data Strcutures uses:
The data structures were used in the dynamic programming task scheduler are heaps, classes, and tables.

1. Heaps are implemented to sort the tasks based on the priority that they have. This allows us to sort them easily as heaps are an effective and efficient way of sorting tasks as compared methods such as like sorting. They also makes retrieval of tasks much easier as we just need to extract the root node in order to get the task with the highest priority.

2. Classes: three classes have been implemented. 
- MaxHeap: class that has the attributes and methods needed for the heap. 
- Task class: takes the list input and compiles them into a formate we can work with in our schedular. It only has attributes and no functionality as it is just used for converting the input into the required formate.
- TaskScheduler class: has the main functionality of the algorithm in its methods. 

The Task and the Task Schedular class created separately because they offer easy maintenance of the code. Changes to the TaskScheduler class do not affect the way the input is given and the Task class makes it easier to change the way a certain input is percieved by the algorithm which was very useful especially when two different implementations of the greedy algorithm approach were done as it allowed versatility of approaches with respect to the utility value.

3. Table were used to get the best possible value of profit. The table looks at all the combinations of tasks calculates the profit of each combination, comparing them to get a final value that is the highest value in the bottom right position of the table.

#### Assumptions:
In dynamic programming, certain assumptions are made. The first of these is that the timeline for the tasks performed is not defined. This is efficient because it uplifts the constraints of scheduling tasks in a particular order or at a specific time, as observed in certain fixed cases of greedy algorithms. This helps the algorithm decide the best course of action for the intended user.

We also make the assumption is that there is no multitasking between the tasks because if the users tend to multitask then the algorithm might output different results than the calculated ones because the users might or might not be efficient at multitasking and therefore it can go either way from the maximum possible output value. This assumption stands because, while dynamic programming gives the most optimal solutions, it does not add solutions together or run them at the same time, which would be the in silico equivalent of multitasking. 

Another assumption present here is that the user who has decided the activities that need to be put in has an elaborate knowledge of the costs and benefits of each activity. A cost could be defined as the amount of money spent on an activity. For example, using a public court for playing basketball has no negative cost because the court is free and within walking range of the residence. However, going to a shopping mall and buying goods will have a negative cost in terms of money spent because the goods are exchanged for money which decreases the bank balance. Since there is an element of happiness in going shopping and buying the goods that are necessary, it can be thought of as a payoff between the amount spent on the goods and the procurement of necessary goods and the happiness associated with it. We can use these to decide whether the gain outweighs the expenditure.

In [175]:
class Task_dp:
    """
    - id: Task Id   
    - description: Short description of the task   
    - duration: Duration in minutes   
    - dependencies: An array of id's of tasks that the task depends on 
    - status: Current status of the task
    - multi_tasking: Tells whether the task can be multitasked or not
    - fixed: Bool for if the task has a fixed start time or can start at any time within its urgency
    - priority: The urgency of the task 
   
    """
    #Initializes an instance of Task, since I am using the same code for Q4, I just set the multitasking part to be
    #default false in this case so that we do not need to account for it in this question.
    #the default priority I am using is 1440 which is equal to the number of minutes in a day
    #fixed is the attribute used to account for the case that there is a task that needs to be done at a certain time
    #in the day
    def __init__(self,task_id,description,duration,dependencies, priority, cost = 0, multi_tasking=False, status="N"):
        self.id = task_id
        self.description = description
        self.duration = duration
        self.dependencies = dependencies
        self.status = status
        self.cost = cost
        self.priority = priority - cost
        self.multi_tasking = multi_tasking
    
    #The format of the display
    def __repr__(self):
        return f"{self.description} - id: {self.id}\n \tDuration:{self.duration}\n\tDepends on: {self.dependencies}\n\tMust be completed at most {self.priority} minutes after day start\n\tFixed task start: {self.fixed}\n\tStatus: {self.status}"
    
    #Definining the comparison variable which in this case is the priority
    def __lt__(self, other):
        return self.priority < other.priority


In [176]:

class DP_MultiTaskScheduler:
    """
    A Daily Task Scheduler Using Priority Queues based on priority that allows for multi-tasking
    Attributes
    ----------
    tasks : arr
        A Python list of all the tasks we schedule
    priority_queue: arr
        The priority queue of tasks ready to execute
    fixed_tasks: arr
        The priority queue of tasks that start at a fixed time
    """ 
    
    NOT_STARTED ='N'
    IN_PRIORITY_QUEUE = 'I'
    COMPLETED = 'C'
    
    def __init__(self, tasks):
        self.tasks = tasks
        self.priority_queue = MaxHeapq()
                
    def print_self(self):
        print('Input List of Tasks')
        for t in self.tasks:
            print(t)            
            
    def remove_dependency(self, task_id):
        """
        Input: list of tasks and id of the task completed
        Output: lists of tasks with completed task removed
        """
        for t in self.tasks:
            if t.id != task_id and task_id in t.dependencies:
                t.dependencies.remove(task_id)           
            
    def get_tasks_ready(self):
        """ 
        Input: list of tasks
        Output: list of tasks that are executable right now(no dependencies)
        """
        for task in self.tasks:
            if task.status == self.NOT_STARTED and not task.dependencies: 
                task.status = self.IN_PRIORITY_QUEUE 
                self.priority_queue.heappush(task)
    
#    def update_priority(self, time_elapsed=0):
        """
        Input: duration of the most recent executed task
        Output: boolean (checks the status of all tasks and returns True if at least one task has status = 'N'
        """
#        for i in range(len(self.tasks)): 
#            self.tasks[i].priority -= time_elapsed
        
    def check_unscheduled_tasks(self):
        """
        Input: list of tasks 
        Output: boolean (checks the status of all tasks and returns True if at least one task has status = 'N'
        """
        for task in self.tasks:
            if task.status == self.NOT_STARTED:
                return True
        return False   
    
    def format_time(self, time):
        return f"{time//60}h{time%60:02d}"
    
    #method that checks for and executes multi-tasking
    
    def get_combinations_ready(self):
        times = []
        profits = []
        for task in self.tasks:
            times.append(task.duration)
            profits.append(task.priority)
        
        return times, profits
    
    def best_possible_combination(self, time = 960):
        times = self.get_combinations_ready()[0]
        profits = self.get_combinations_ready()[1]
        n = len(profits)
        K = [[0 for x in range(time + 1)] for x in range(n + 1)]
        
        for i in range(n + 1):
            for w in range(time + 1):
                if i == 0 or w == 0:
                    K[i][w] = 0
                elif times[i - 1] <= w:
                    K[i][w] = max(profits[i - 1] + K[i - 1][w - times[i - 1]], K[i - 1][w])
                else:
                    K[i][w] = K[i - 1][w]
                    
        return K[n][w]

    def multi_tasker(self, task, current_time, profit):
        if (task.multi_tasking):
            temporary = []
            while(self.priority_queue.heap_size > 0 and (not self.priority_queue.maxk().multi_tasking or (self.priority_queue.maxk().duration))):
                temporary.append(self.priority_queue.heappop())

            # if there are tasks in the priority queue after the while loop is done, we can multi-task using them
            if self.priority_queue.heap:
                multitask = self.priority_queue.heappop()
                if ((current_time + task.duration)//60 < 24) and ((current_time + multitask.duration)//60 < 24):
                    print(f"⏰Simple Scheduler at time {self.format_time(current_time)} started executing tasks {task.id} and {multitask.id} that take {task.duration} and {multitask.duration} mins respectively")
    
                #Increment the current time with the duration of the larger task(duration wise) as that task will still
                #be in process when the first task is completed.
                if task.duration < multitask.duration:
                    current_time += task.duration
                    if ((current_time + task.duration)//60 < 24) and ((current_time + multitask.duration)//60 < 24):
                        profit += task.priority + multi_task.priority 
                        print(f"✅ Completed Task {task.id} - '{task.description}' at time {self.format_time(current_time)}\n")
                        current_time += multitask.duration - task.duration
                        print(f"✅ Completed Task {multitask.id} - '{multitask.description}' at time {self.format_time(current_time)}\n")
        
                else:
                    current_time += multitask.duration
                    if ((current_time + task.duration)//60 < 24) and ((current_time + multitask.duration)//60 < 24):
                        profit += task.priority + multi_task.priority
                        print(f"✅ Completed Task {multitask.id} - '{multitask.description}' at time {self.format_time(current_time)}\n")
                        current_time += task.duration - multitask.duration
                        print(f"✅ Completed Task {task.id} - '{task.description}' at time {self.format_time(current_time)}\n")
                        
                self.remove_dependency(task.id)
                self.remove_dependency(multitask.id)
                task.status = self.COMPLETED
                multitask.status = self.COMPLETED

            #if the priority queue is empty after the while loop:
            #none of the tasks can be multitasked 
            #within the time between the current time and the start of the next fixed task
            #so we just perform the initial task without multi-tasking
            else:
                if (current_time + task.duration)//60 < 24:
                    print(f"⏰Simple Scheduler at time {self.format_time(current_time)} started executing task {task.id} that takes {task.duration} mins")
                    current_time += task.duration
                    print(f"✅ Completed Task {task.id} - '{task.description}' at time {self.format_time(current_time)}\n")  
                    profit += task.priority
                self.remove_dependency(task.id)
                task.status = self.COMPLETED

            #once we've carried out the task:
            #we move all the other tasks to the priority queue from the temporary list we had added them to
            for i in temporary:
                self.priority_queue.heappush(i)
            
        #if the task isn't multi-taskable we treat it as a regular task without running it through further comparisons
        else: 
            if (current_time + task.duration)//60 < 24:
                print(f"⏰Simple Scheduler at time {self.format_time(current_time)} started executing task {task.id} that takes {task.duration} mins")
                current_time += task.duration
                print(f"✅ Completed Task {task.id} - '{task.description}' at time {self.format_time(current_time)}\n")
                profit += task.priority
            self.remove_dependency(task.id)
            task.status = self.COMPLETED
           
        #return the current time
        return current_time, profit
    
    
    
    def run_task_scheduler(self, starting_time = 480):
        print(f"Maximum possible profit = {self.best_possible_combination()}")
        current_time = starting_time
        profit = 0
        while self.check_unscheduled_tasks() or self.priority_queue.heap:
            self.get_tasks_ready()
            if current_time//60 > 24:
                print("End of the day. No more tasks possible today")
                break
            if self.priority_queue.heap_size > 0:  
                if self.priority_queue.maxk().duration:  
                    temporary = []
                    while(self.priority_queue.heap_size > 0):
                        temporary.append(self.priority_queue.heappop())   
                    task = temporary.pop(0)  
                        
                        # calling the multi_task method and update the current time to the 
                        #new time once the multi-task method is done.
                    current_time, profit = self.multi_tasker(task, current_time, profit)
        
                    for i in temporary:
                        self.priority_queue.heappush(i)
                        
                #if the current task can be done before the next fixed task but can not be multi-tasked with another
                #task, we just perform it like a normal task.
                else:
                    task = self.priority_queue.heappop()    
                    current_time, profit = self.multi_tasker(task, current_time, profit)
        
        #profit = self.multi_tasker(task, current_time, profit)[1]             
        total_time = current_time - starting_time             
        print(f"🏁 Completed all planned tasks in {total_time//60}h{total_time%60:02d}min")
        print(f"Profit achieved through this combination of activities = {profit}")
    
    
    

In [177]:
task_list = [
    Task_dp(0, 'waking up', 10, [], 100),
    Task_dp(1, 'having breakfast', 20, [0], 75, multi_tasking=True),
    Task_dp(2, 'watch an episode of Community', 30, [], 35, 5, multi_tasking = True),
    Task_dp(3, 'take a walk', 30, [], 25, 10, multi_tasking = True),
    Task_dp(4, 'Do pre-class work', 180, [] , 80),
    Task_dp(5, 'Take class', 90, [4], 90),
    Task_dp(6, 'Go to a park', 180, [], 50, 5, multi_tasking = True),
    Task_dp(7, 'take a trip to the beach', 400, [], 30, 60),
    Task_dp(8, 'Attend a conference', 300, [], 55, 5),
    Task_dp(9, 'get groceries', 50, [], 70, 30),
    Task_dp(10, 'make dinner', 40, [9], 80, 5),
    Task_dp(11, 'socialize', 300, [], 60, 5)]


task_scheduler = DP_MultiTaskScheduler(task_list)
task_scheduler.run_task_scheduler()

Maximum possible profit = 605
⏰Simple Scheduler at time 8h00 started executing task 0 that takes 10 mins
✅ Completed Task 0 - 'waking up' at time 8h10

⏰Simple Scheduler at time 8h10 started executing task 4 that takes 180 mins
✅ Completed Task 4 - 'Do pre-class work' at time 11h10

⏰Simple Scheduler at time 11h10 started executing task 5 that takes 90 mins
✅ Completed Task 5 - 'Take class' at time 12h40

⏰Simple Scheduler at time 12h40 started executing task 1 that takes 20 mins
✅ Completed Task 1 - 'having breakfast' at time 13h00

⏰Simple Scheduler at time 13h00 started executing task 11 that takes 300 mins
✅ Completed Task 11 - 'socialize' at time 18h00

⏰Simple Scheduler at time 18h00 started executing task 8 that takes 300 mins
✅ Completed Task 8 - 'Attend a conference' at time 23h00

⏰Simple Scheduler at time 23h00 started executing task 9 that takes 50 mins
✅ Completed Task 9 - 'get groceries' at time 23h50

🏁 Completed all planned tasks in 15h50min
Profit achieved through this

In [178]:
task_list_x = [
    Task_dp(0, 'A', 200, [], 100),
    Task_dp(1, 'B', 200, [0], 75),
    Task_dp(2, 'C', 300, [], 35, 5),
    Task_dp(3, 'D', 300, [], 25, 10),
    Task_dp(4, 'E', 180, [] , 80),
    Task_dp(5, 'F', 900, [4], 90),
    Task_dp(6, 'G', 180, [], 50, 5)]

task_scheduler = DP_MultiTaskScheduler(task_list_x)
task_scheduler.run_task_scheduler()

Maximum possible profit = 300
⏰Simple Scheduler at time 8h00 started executing task 0 that takes 200 mins
✅ Completed Task 0 - 'A' at time 11h20

⏰Simple Scheduler at time 11h20 started executing task 4 that takes 180 mins
✅ Completed Task 4 - 'E' at time 14h20

⏰Simple Scheduler at time 14h20 started executing task 1 that takes 200 mins
✅ Completed Task 1 - 'B' at time 17h40

⏰Simple Scheduler at time 17h40 started executing task 6 that takes 180 mins
✅ Completed Task 6 - 'G' at time 20h40

🏁 Completed all planned tasks in 12h40min
Profit achieved through this combination of activities = 300


#### Analysis of test case 1

The above example of a schedular hihglights an important part of the algorithm: its upper bound. This dynamic programming task schedular takes into account time constraints that exist in the real world; all the tasks cannot be scheduled within a limited amount of time and hence requires an optimal combination that can give us the highest utility in n number of hours. The value of n can be changed by the user depending on how much time they have in a day but here n was taken as a baseline value corresponding with the mathematical end of the day: 1200 am. 


In [180]:
#two test cases 

task_list_y = [
    Task_dp(0, 'A', 100, [], 100),
    Task_dp(1, 'B', 20, [0], 75),
    Task_dp(2, 'C', 300, [], 35, 5),
    Task_dp(3, 'D', 30, [], 25, 10)]

task_list_z = [
    Task_dp(0, 'A', 100, [], 100),
    Task_dp(1, 'B', 20, [0], 75),
    Task_dp(2, 'C', 300, [], 35, 5),
    Task_dp(3, 'D', 30, [], 25, 10),
    Task_dp(4, 'E', 180, [] , 30, 60)]


print("This is an example of a DP task schedular not including a task with a negative utility")
task_scheduler_y = DP_MultiTaskScheduler(task_list_y)
task_scheduler_y.run_task_scheduler()

print('-------------------------------------------------------------------------')

print("This is an example of a DP task schedular including a task with a negative utility")
task_scheduler_z = DP_MultiTaskScheduler(task_list_z)
task_scheduler_z.run_task_scheduler()

This is an example of a DP task schedular not including a task with a negative utility
Maximum possible profit = 220
⏰Simple Scheduler at time 8h00 started executing task 0 that takes 100 mins
✅ Completed Task 0 - 'A' at time 9h40

⏰Simple Scheduler at time 9h40 started executing task 1 that takes 20 mins
✅ Completed Task 1 - 'B' at time 10h00

⏰Simple Scheduler at time 10h00 started executing task 2 that takes 300 mins
✅ Completed Task 2 - 'C' at time 15h00

⏰Simple Scheduler at time 15h00 started executing task 3 that takes 30 mins
✅ Completed Task 3 - 'D' at time 15h30

🏁 Completed all planned tasks in 7h30min
Profit achieved through this combination of activities = 220
-------------------------------------------------------------------------
This is an example of a DP task schedular including a task with a negative utility
Maximum possible profit = 220
⏰Simple Scheduler at time 8h00 started executing task 0 that takes 100 mins
✅ Completed Task 0 - 'A' at time 9h40

⏰Simple Schedule

#### Analysis of test case 2

The above test case shows a very integral part of dynammic programming: its ability to calculate the optimal combination of task that can give the highest aggregate utility. In the first example, a list of tasks are given that have the utility of 220. The output of the schedular corresponds with this. But in the second example, a task with a negative utility is added:

priority - cost = 30 - 60 = -30

The output of the second example still adds the task in the schedule because we can still fit more tasks in the day, but the utility value of the best possible combination does not match the utility value of the schedular. 

### Time and Space Complexity


##### Time Complexity
A problem that I faced when trying to graphically compute the time compelxity of these task schedulars is that it is very hard to generate an imput by hand even and even with a random generation. Hence, we rely on time recurrence relations with respect to the logical reasoning of how the algorithm works.
Time recurrence calculation

Cost the algorithm:

Best_possible_combination() = O(n^2) as O(n) for the rows and O(n) for the columns which make it equal to O(n^2)

Multi_tasker() = O(n) x O(n - 1) 
The function is runs on all the tasks once and then once again without the task that was performed. This simplifies down to O(n^2) - O(n).

Run_taskscheduler() = O(n) x O(n - 1) 
runs the task and then the method again on the tasks that are left. Can be simplified as O(n^2) - O(n).

The methods within the class that are called such as get_tasks_ready(), get_combinations_ready(), and get_unscheduled_tasks() take O(n), which can be written as C*O(n) when performed C number of times. 

The runtime of the heappush() method is O(n log n) 
heappush = n
insertion = log n.

The extraction of the root node takes O(1) time and multiplied by a constant means the total time will be K*O(1)

Thus, the time recurrence equation becomes:

T(n) = O(n^2) + O(n^2) - O(n) + O(n^2) - O(n) + CO(n) + O(n log n) + KO(1)

The 2O(n) + CO(n) can be written as C*O(n) as C is just a constant

T(n) = 3 O(n^2) + O(n log n) + CO(n) + KO(1)

For a very large input size, only the O(n^2) will be important and the others can be disregarded as their growth rate is too small for significant impact.

Thus, the final time recurrence equation becomes:

T(n) = O(n^2)

This means that for each additional task added to the task scheduler, the time required to run the algorithm will grow exponentially over time. This also makes sense in the sense of the intuitive reasoning as each new task will bring in a whole row and column to the table we are using which makes the growth "n^2"

##### Space Complexity¶

For space complexity, the dynamic algorithmic approach requires more space as compared to the greedy algorithm one because it makes use of the table and thus has to store all the values in the memory. Hence, when comparing space required for each approach, dynamic would require more space.


# 3. Analysis of Schedulars 2.0

Things to Consider:

a. Do all of these implementations make use of the same data structures?
These implementation make use of heaps and Classes, similar to how we did in the simple Schedular, but in addition to that dynamic programming also ues a table to calculate the highest profit value associated with the tasks. This adds to its space and time complexity but is also helpful in going through all the possible combinations. 

b. Do they address all the concerns raised in question 1?
The two algorithms take into account the concerns of the simpel schedular. They do the following: 
- Have an upper bound, n, beyond which more tasks cannot be done in a day
- tasks care scheduled according to their priorities 
- tasks have profits are costs associated with them that reflect their perfecieved utility 
- more tasks can be added to the input and it will be scheudled according to its priority, hence it is flexible and applicable.

c. What are the metrics that you are using to compare the schedule and the scheduler efficiencies?
The schedule is considered to be efficient if it gives us the highest aggregate profit value. It is also considered through its use in real life, how well it can capture the needs to different users and how realistically it can represent the schedule of an ordinary user. 

# HC and LO Appendix 

#### LOs used:
#algorithmicStrategies: Two different algorithms are used to approach the same problem of task schedular. I explain the specific principles of both greedy algorithm and dynamic programming to justify their role in creating an optimal schedular. Moreover, based on these algorithmic principles I also show what kinds of inputs work in its favour and how it accomplishes the goal of optimizing scheduling.  

#computationalCritique: I critigue my previous schedular and analyze the limitations of the algorithm I used for it and its inability to effectively capture the needs of the problem. 

#ComplexityAnalysis: I analyze greedy algorithm and dynamic programming based on their time and space compelxiy. I justify why I dont use a graphical or computational approach and analyze the compelxity through recurrance equation instead, breaking it down and giving an equation for each part of the algorithm. 


#### HCs used: 
#utility: utility is used to calculate the priority values of the tasks. A cost benefit analysis is done for each task before its perceived utility is given to the tasks. This allows us to make our algorithm applicable

#critque: I analyze the limitations of my previous schedular and based on those limitations I create aspects in the new algorithms that can counter them. Hence, combining analysis with action. 

#optimization: The objective function here was to make a schedular that outputs a list of tasks with the highest profit associated with them. I created schedular with two different approaches to optimize the profits. 
